### PR TITLE
Update runtime to 5.15-21.08

### PR DESCRIPTION
**Pull request by travier**
The 5.15 version of the KDE Runtime is based on the 20.08 version of the Freedesktop Runtime and will stay as-is to keep compatibility.
The 5.15-21.08 version of the KDE Runtime is based on the 21.08 Freedesktop one. The Qt/KDE Libraries are mostyl similar between the two runtimes.

This change is mostly maintenance to keep the base runtime updated.

### CODE REVIEW

**Comment by mhoeher** [comment]
Hi @travier,

thanks for this PR! One question: Is there somewhere some kind of documentation on which version of the runtime to use? I was actually not aware at all about this variant of the runtime at all :open_mouth:

**Comment by hfiguiere** [comment]
the announcement was late but we hope it'll be right next time

https://discourse.flathub.org/c/announcements/6

**Comment by mhoeher** [comment]
Hi @hfiguiere,

thanks for the link! Interesting to see there's already a runtime with Qt 6 :smiley:

**Comment by mhoeher** [comment]
Hi @travier and @hfiguiere,

after migrating to the new runtime, a user reported issues with the app on ARM64, see https://gitlab.com/rpdev/opentodolist/-/issues/503

Just out of curiosity: Did you experience similar issues with other apps as well? The new runtime seems to work just fine on x86_64 (this is also where I tested the app). Unfortunately, I do not have an ARM64 device available and reproducing in a VM is a bit difficult (at least I am not sure if the issues I see in the VM I just set up are due to the bug reported or due to the VM setup itself...)

If you'd have any hints, I would be more than thankful!

In the meantime, I reverted back to the previous runtime, which used to work.

**Comment by hfiguiere** [comment]
runs fine here on my RaspberryPi 4.

**Comment by mhoeher** [comment]
Hi @hfiguiere,

thanks for testing! Just to make sure: Did you use the latest stable version (this one I reverted back to the previous runtime) or the beta one (this one is still on the new runtime)?

**Comment by hfiguiere** [comment]
the one from stable.

**Comment by hfiguiere** [comment]
beta crashes deep in the bowels of QtQml. the first 15 stack traces are in there.

**Comment by mhoeher** [comment]
Hi @hfiguiere,

first of all: Happy new new to you! And thanks a lot for cross checking.

That's bad (and a bit what I was afraid of). Seems there's some difference in behaviour between the x86_64 and ARM64 versions of Qt in this case which cause the crash. Are you by chance aware of other apps that suffered the same symptoms?

**Comment by travier** [comment]
Sorry, I don't have aarch64 hardware to try.

**Comment by mhoeher** [comment]
Hi @travier,

no problem, I have the same issue (no hardware I could run and debug this on readily at hand).

What I found is this: https://people.gnome.org/~alexl/presentations/guadec-2018-workshop.pdf

So it seems I can install the ARM version of the app on an x86_64 host and - having `qemu-user-static` installed - also run it (and well, it crashes, as expected). However, unfortunately spawning a shell in the ARM sandbox using

```bash
flatpak run  --devel --arch=aarch64 --command=sh net.rpdev.OpenTodoList/aarch64/beta
```
and trying to run `gdb` on the app does not work. However, I can run the app itself (which at least yields a bit of output via `abrt`). Running the app yields a bus error and looking at the stack trace in `abrt` shows that the main trace seemed to crash in a function called `__memcpy_avx_unaligned_erms`. I checked the Qt bug tracker, and the only issue found seems to be this one: https://bugreports.qt.io/browse/QTBUG-80944 - however, I am not sure if this is related at all. In addition, I am not sure if the stack trace shown in `abrt` is reliable, since in my case I have the emulation in between, and the user reporting the issue on GitLab already mentioned they saw a `SIGSEGV` and not a `SIGBUS`.

**Comment by mhoeher** [comment]
I just spent some minutes thinking about the issue again... What I did:

First, I ran the ARM version of the app (via `qemu-user-static`). The app crashes after producing the following log:

```
Cyclic dependency detected between "qrc:/Utils/ItemUtils.qml" and "qrc:/Utils/SyntaxHighlighter.qml"
```

Afterwards, I ran the x86_64 version, which also has this log line but continues a bit:

```
Cyclic dependency detected between "qrc:/Utils/ItemUtils.qml" and "qrc:/Utils/SyntaxHighlighter.qml"
Cyclic dependency detected between "qrc:/Utils/ItemUtils.qml" and "qrc:/Controls/TextField.qml"
Cyclic dependency detected between "qrc:/Utils/ItemUtils.qml" and "qrc:/Windows/DeleteItemDialog.qml"
Cyclic dependency detected between "qrc:/Utils/ItemUtils.qml" and "qrc:/Controls/Label.qml"
Cyclic dependency detected between "qrc:/Utils/ItemUtils.qml" and "qrc:/Windows/DeleteCompletedItemsDialog.qml"
Cyclic dependency detected between "qrc:/Utils/ItemUtils.qml" and "qrc:/Windows/DateSelectionDialog.qml"
```

So I somehow suspect the error might be in the file `SyntaxHighlighter.qml`. This file in turn is quite simple:

```qml
pragma Singleton

import QtQuick 2.0
import OpenTodoList 1.0 as OTL
import "." as Utils

OTL.SyntaxHighlighter {
    theme: Utils.Colors.usingDarkColorTheme ? OTL.SyntaxHighlighter.Dark : OTL.SyntaxHighlighter.Light
}
```

Basically, it instantiates a `SyntaxHighlighter` object from the C++ part of the app: https://gitlab.com/rpdev/opentodolist/-/blob/stable/lib/utils/syntaxhighlighting.h

This class is a simple wrapper around KDE framework's SyntaxHighlighting library. Long story short: I remember that deeply inside, that library uses `QJsonDocument::fromBinary` (see [here](https://doc.qt.io/qt-5/qjsondocument-obsolete.html#fromBinaryData)). IIRC, this API has some issues with alignment on certain platforms (in fact, I had issues with it on Android ARM in the past) - which would potentially explain bus errors.

@hfiguiere: You mentioned you've got some back traces? Could you share them here, please?

In general, does the above (SyntaxHighlighting module and crashes on ARM platform) ring some bells?

**Comment by mhoeher** [comment]
Hi @hfiguiere and @travier,

sorry for spamming you, but maybe this is relevant for you as well (as this might affect other apps, too):

I got a backtrace for the issue (see [here](https://gitlab.com/rpdev/opentodolist/-/issues/503#note_799535933)). In fact, the backtrace does point to a problem (probably) entirely in Qt/QML.

As I have completely no clue on how to continue on this issue, I reported a bug in the Qt bugtracker: [QTBUG-99545](https://bugreports.qt.io/browse/QTBUG-99545). Maybe the Qt devs can shed some light on this. If you have any additional information, maybe you could also comment in the Qt bug.

**Comment by deepbluev7** [comment]
> Hi @travier and @hfiguiere,
> 
> after migrating to the new runtime, a user reported issues with the app on ARM64, see https://gitlab.com/rpdev/opentodolist/-/issues/503
> 
> Just out of curiosity: Did you experience similar issues with other apps as well? The new runtime seems to work just fine on x86_64 (this is also where I tested the app). Unfortunately, I do not have an ARM64 device available and reproducing in a VM is a bit difficult (at least I am not sure if the issues I see in the VM I just set up are due to the bug reported or due to the VM setup itself...)
> 
> If you'd have any hints, I would be more than thankful!
> 
> In the meantime, I reverted back to the previous runtime, which used to work.

Nheko did actually experience similar issues when upgrading the SDK and as such we immediately reverted that change and I didn't find time to debug it yet. Our backtraces looked very similar to yours: https://github.com/Nheko-Reborn/nheko/issues/816#issuecomment-979453151 (There are also some other bugs in that issue, so don't get confused.) So yeah, just wanted to confirm that other apps also experienced that issue.

**Comment by mhoeher** [comment]
Hi @deepbluev7,

thanks for letting me know! I'll also add this to the Qt issue for reference.

**Comment by mhoeher** [comment]
I received some feedback on the Qt bug tracker:

It seems that there's one particular change between the previous Flatpak runtime and the new updated one, see my comment here: https://bugreports.qt.io/browse/QTBUG-99545?focusedCommentId=639230&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-639230

The first idea would be to revert this change and see if the issue still occurs. As I am not experienced with building these Flatpak runtimes: Can you guys advice on how to best proceed on this?

I used to compile the QtRemoteObjects module within my app some while ago (as it was - for a long time - not part of the runtime). However, that was quite challenging (as Flatpak does not allow installing into the "system" locations). My naive thinking would suggest that for testing one could try to do the same (i.e. build the old version of QtDeclarative within one of the affected apps and see if the crash still occurs). However, I am not sure if this will be easy: For QtRO, I added a completely new module, while in this case I would have to somehow replace/override the one already present in the runtime - I guess this will cause some headaches. So, any advice here?

**Comment by travier** [comment]
Ping @aleixpol (patch author)

**Comment by aleixpol** [comment]
Commented in the Qt bug report.

**Comment by hfiguiere** [comment]
> The first idea would be to revert this change and see if the issue still occurs. As I am not experienced with building these Flatpak runtimes: Can you guys advice on how to best proceed on this?

I have done that, but last time it took over 24 hours. Not sure if we can do a PR like a regular flathub package, which could be easier / faster.

It seems that the same problem is being experienced on https://github.com/flathub/com.cuperino.qprompt/pull/1

**Comment by hfiguiere** [comment]
The latest update of the runtime fixees the problem, at least work around by not using the build optimisation option that cause the problem. It didn't crash here, no rebuild necessary.
